File: examples/monorepo_example/rails/app/controllers/jobs_controller.rb
class JobsController < ApplicationController
  def enqueue
    job_type = params[:job_type] || "example"
    
    case job_type
    when "example"
      job = ExampleJob.perform_later(message: params[:message] || "Hello from Rails!")
    when "elixir"
      job = ElixirProcessedJob.perform_later(
        user_id: params[:user_id] || 123,
        action: params[:action] || "process"
      )
    when "zig"
      job = ZigProcessedJob.perform_later(
        message: params[:message] || "Hello from Rails to Zig!"
      )
    when "zig_example"
      job = ZigExampleJob.perform_later(
        message: params[:message] || "Example job for Zig"
      )
    when "globalid"
      # Test GlobalID resolution
      user = User.find(rand(1..100))
      job = GlobalidTestJob.perform_later(
        user: user,
        message: "GlobalID test from Rails UI at #{Time.current}"
      )
    when "concurrency"
      # Test concurrency limits
      resource_id = params[:resource_id] || rand(1..10).to_s
      job = ConcurrencyTestJob.perform_later(
        key: resource_id,
        message: "Concurrency test from Rails UI at #{Time.current}"
      )
    when "cross_language_concurrency"
      # Test cross-language concurrency
      resource_id = params[:resource_id] || rand(1..5).to_s
      job = CrossLanguageConcurrencyJob.perform_later(
        resource_id: resource_id,
        action: "process"
      )
    else
      if request.format.html?
        redirect_to root_path, alert: "Unknown job type: #{job_type}"
      else
        render json: { error: "Unknown job type" }, status: :bad_request
      end
      return
    end
    
    # Broadcast job update
    ActionCable.server.broadcast("jobs", { type: "job_enqueued", job_type: job_type })
    
    if request.format.html?
      redirect_to root_path, notice: "Job enqueued successfully: #{job_type}"
    else
      render json: { status: "enqueued", job_type: job_type }
    end
  end
end

File: examples/monorepo_example/elixir/lib/monorepo_example_web/controllers/jobs_controller.ex
defmodule MonorepoExampleWeb.JobsController do
  @moduledoc """
  Controller for handling job enqueueing requests.
  """
  use MonorepoExampleWeb, :controller

  def enqueue(conn, %{"job_type" => "elixir"}) do
    # Enqueue an Elixir job
    case MonorepoExample.Jobs.ScheduledElixirJob.perform_later(%{
           message: "Enqueued from Elixir web interface at #{DateTime.utc_now()}"
         }) do
      {:ok, _job} ->
        conn
        |> put_flash(:info, "Elixir job enqueued successfully!")
        |> redirect(to: "/")

      {:error, reason} ->
        conn
        |> put_flash(:error, "Failed to enqueue job: #{inspect(reason)}")
        |> redirect(to: "/")
    end
  end

  def enqueue(conn, %{"job_type" => "ruby"}) do
    # Enqueue a Ruby job using the descriptor module
    case MonorepoExample.Jobs.ExampleRubyJob.perform_later(%{
           message: "Enqueued from Elixir web interface at #{DateTime.utc_now()}"
         }) do
      {:ok, _job} ->
        conn
        |> put_flash(:info, "Ruby job enqueued successfully!")
        |> redirect(to: "/")

      {:error, reason} ->
        conn
        |> put_flash(:error, "Failed to enqueue Ruby job: #{inspect(reason)}")
        |> redirect(to: "/")
    end
  end

  def enqueue(conn, %{"job_type" => "zig"}) do
    # Enqueue a Zig job - will be processed by Zig worker
    case GoodJob.enqueue(MonorepoExample.Jobs.ZigJob, %{
           message: "Enqueued from Elixir web interface at #{DateTime.utc_now()}"
         }) do
      {:ok, _job} ->
        conn
        |> put_flash(:info, "Zig job enqueued successfully!")
        |> redirect(to: "/")

      {:error, reason} ->
        conn
        |> put_flash(:error, "Failed to enqueue Zig job: #{inspect(reason)}")
        |> redirect(to: "/")
    end
  end

  def enqueue(conn, %{"job_type" => "globalid"}) do
    # Test GlobalID resolution - enqueue a job with a GlobalID
    # This simulates what Rails would send
    case GoodJob.enqueue("GlobalIDTestJob", %{
           user: %{
             "_aj_globalid" => "gid://myapp/User/#{Enum.random(1..100)}"
           },
           message: "GlobalID test from Elixir UI at #{DateTime.utc_now()}"
         }) do
      {:ok, _job} ->
        conn
        |> put_flash(:info, "GlobalID test job enqueued successfully!")
        |> redirect(to: "/")

      {:error, reason} ->
        conn
        |> put_flash(:error, "Failed to enqueue GlobalID test job: #{inspect(reason)}")
        |> redirect(to: "/")
    end
  end

  def enqueue(conn, %{"job_type" => "concurrency"}) do
    # Test concurrency limits
    resource_id = Enum.random(1..10) |> to_string()
    case GoodJob.enqueue("ConcurrencyTestJob", %{
           key: resource_id,
           message: "Concurrency test from Elixir UI at #{DateTime.utc_now()}"
         }, concurrency_key: resource_id) do
      {:ok, _job} ->
        conn
        |> put_flash(:info, "Concurrency test job enqueued successfully!")
        |> redirect(to: "/")

      {:error, reason} ->
        conn
        |> put_flash(:error, "Failed to enqueue concurrency test job: #{inspect(reason)}")
        |> redirect(to: "/")
    end
  end

  def enqueue(conn, %{"job_type" => "cross_language_concurrency"}) do
    # Test cross-language concurrency
    resource_id = Enum.random(1..5) |> to_string()
    case GoodJob.enqueue("CrossLanguageConcurrencyJob", %{
           resource_id: resource_id,
           action: "process"
         }, concurrency_key: "resource:#{resource_id}") do
      {:ok, _job} ->
        conn
        |> put_flash(:info, "Cross-language concurrency test job enqueued successfully!")
        |> redirect(to: "/")

      {:error, reason} ->
        conn
        |> put_flash(:error, "Failed to enqueue cross-language concurrency test job: #{inspect(reason)}")
        |> redirect(to: "/")
    end
  end

  def enqueue(conn, _params) do
    conn
    |> put_flash(:error, "Invalid job type")
    |> redirect(to: "/")
  end
end
